added code to enable cacheing of the gists

kiran3807 opened the pull request:
Right now we can display the  different part of a gist on the basis of line numbers . We can do it by writing a single gist and using multiple `<code>` elements that display different parts of the gist. However that causes the same gist to be loaded multiple times leading to unnecessary network calls. I have simply added an option `data-gist-enable-cache`which when set to true, caches the calls to gists with the same id. For example 

 `<code data-gist-id="4567" data-gist-enable-cache="true"></code>` 

will cause all the `<code>` elements with gists which refer to the id 4567 to be cached.

As of now the order where the option `data-gist-enable-cache` is set to true, matters. That is the only calls that are made by `<code>` elements with same gist ids after the element where the option is set true, will be cached.

If cacheing is cool then another option might be to pass an object to the` gist()` function as an argument to enable/disabling cacheing.

Thanks :)

blairvanderhoof commented on gist-embed.js:
```diff
@@ -182,6 +191,35 @@
+                gistCache.put(url,$.Deferred());    
+            }
+        },
+        cache : true,
```
Isn't this a separate jQuery cache? Do we need both local and this one?

kiran3807 replied:
Oh this is my bad, I had left it there by mistake while I tried using Jquery's cache  for achieving the aim. Somehow it did not workout, thats why I used a separate local cache. will remove it when I submit the final changes after the discussion  in the other note is completed. :)

blairvanderhoof commented on gist-embed.js:
```diff
@@ -182,6 +191,35 @@
+      }
+      
+      // request the json version of this gist
+      $.ajax({
```
Why do we need to use $.Deferred? 

For a simpler option, we could simply keep a dictionary called cachedQueries = {}.

The key would simply be the gist id.

Instead of doing the logic in the ajax call, simply wrap that in an if/else where the if condition is checking if cachedQueries[id] exists. If so, call successCallback with that value, if not just run the existing ajax query and in success, store the results in cachedQueries[id] = response.

kiran3807 replied:
True, however lets consider this use case :
```
<code  data-gist-id="e85a1d560986dc5c3fa62d8831ed514d" data-gist-enable-cache="true" data-gist-line="12-25"></code>
<code  data-gist-id="e85a1d560986dc5c3fa62d8831ed514d"  data-gist-line="10-12"></code>
<code  data-gist-id="e85a1d560986dc5c3fa62d8831ed514d"  data-gist-line="1-9"></code>
```
Here the requests will be sent one after another. First the request for gist1, Req1 will be sent. but before Req1 is resolved successfully, the request for second gist is sent. Now we can only store the response in the cache once the first request is resolved. Hence Req2 will be fired despite the fact that in the success function of Req1 we would set the cache for the current gist id/url.

So to solve the above problem I store a promise object in the cache when Req1 is fired. so before Req2 is fired we find that the cache has the promise object as a proxy for the eventual response and Req2 can be stopped from firing. Then we can simply render the content when the req1 is resolved and use that for the subsequent gists

blairvanderhoof commented on gist-embed.js:
```diff
@@ -27,6 +27,19 @@
+    has : function(key){
+        return this.store[key] ? true : false;
+    }
+  }
```
Ok we can go with your cache idea, thanks.  I don't think this structure is necessary. Can we just have a var, gistCache = {};

get => gistCache[key];
has => if (gistCache[key])
put => gistCache[key] = bar;

blairvanderhoof commented on gist-embed.js:
```diff
@@ -182,6 +191,35 @@
+        timeout: 20000,
+        beforeSend : function(){
+            // option to enable cacheing of the gists
+            if(enableCache && gistCache.has(url)){
```
Lets reorganize this. Shortcircuit to true if no cache

if (!enableCache) {
  return true;
}

if (gistCache[url]) {

} else {
  gistCache[url] = bar;
}

blairvanderhoof replied:
We can have an outer and inner if

if (enableCache) {
  if (gistCache[url]) {
  } else {
  }
}

blairvanderhoof commented on gist-embed.js:
```diff
@@ -182,6 +191,35 @@
+                    successCallback(response);
+                    if(typeof callback === 'function') callback();
+                },function(error){
+                    //error
```
We should handle the error case. Maybe make an errorCallback so it is reusable here and below.

blairvanderhoof commented on gist-embed.js:
```diff
@@ -182,6 +191,35 @@
+                // loading the response from cache and preventing the ajax call
+                gistCache.get(url).then(function(response){
+                    successCallback(response);
+                    if(typeof callback === 'function') callback();
```
We could make a `completeCallback` reusable here and below.

blairvanderhoof replied:
if (

kiran3807 replied:
Implemented all other suggestions, However I'm not clear regarding making a reusable `completeCallback`. Do you mean I should make a function like this  :

```
function completeCallback(){
      if(typeof callback === 'function') callback();
}
```
and use it where ever I find `if(typeof callback === 'function') callback();` ??

blairvanderhoof commented on gist-embed.js:
```diff
@@ -182,6 +191,35 @@
+            // option to enable cacheing of the gists
+            if(enableCache && gistCache.has(url)){
+                // loading the response from cache and preventing the ajax call
+                gistCache.get(url).then(function(response){
```
Convention for syntax is spaces before a {

e.g. function(response) {

kiran3807 commented:
Hi, 
Implemented the suggestions, namely =>

1. Formatting of `if{ `to `if {`

2. Formating `foo{` to `foo {`

3.added a new function `completeCallBack`

4.refactored `if(a&&b)` to `if(a){` `if(b) }` as mentioned in the specific cases above

5.removed the object `gistCache` and instead replaced it with an array as you suggested. Also modified the relevant conditionals accordingly

blairvanderhoof commented:
Thanks so much @kiran3807 !!

kiran3807 commented:
You'r Welcome :)
